### PR TITLE
Fix catch-all rule handling

### PR DESCRIPTION
**Pull request by un-tone**
For https://github.com/StarcounterApps/Website/issues/58

### CODE REVIEW

File: src/WebsiteProvider/Api/ContentHandlers.cs [hunk 1]
```diff
@@ -93,7 +93,7 @@ public void Register()
 
                 this.CurrentTemplate = template;
                 WrapperPage master = GetLayoutPage();
-                master.IsFinal = webUrl.IsFinal;
+                master.IsFinal = webUrl.IsFinal || string.IsNullOrEmpty(webUrl.Url);
```
Comment (joozek78) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: What if there is no rule that has empty URL?
Also I don't think this change here is good - what if someone legitimately wants a rule that catches all, but isn't final?

Comment (warpech) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Added my comments in https://github.com/StarcounterApps/Website/issues/58#issuecomment-286901166

Comment (un-tone) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Regarding the case when there is no rule that has empty URI, implemented finishing the loop when there is no matching Catching Rule.
Regarding more than one "Catch-all" rules I mention to the 1st constraint https://github.com/StarcounterApps/Website/issues/58#issuecomment-286901166. I just propose to add validation of rules on saving.

**Comment by un-tone** [comment]
Similar https://github.com/StarcounterApps/Website/issues/63 issue still impacts on **Website**  with this PR. Need to be investigated.

**Comment by warpech** [comment]
@un-tone could you pls add a detailed information to README.md about how Catch-all rules work after this change?

**Comment by un-tone** [comment]
> Similar #63 issue still impacts on Website with this PR. Need to be investigated.

Fixed in https://github.com/StarcounterApps/Website/pull/65/commits/b712373da24bac96faba9faaba396c0726c74e37

**Comment by un-tone** [comment]
The issue is fixed. `README.md` is updated. Please review.

File: src/WebsiteProvider/Api/ContentHandlers.cs [hunk 2]
```diff
@@ -132,6 +134,10 @@ protected void RegisterFilter()
                             wrapper = response.Resource as WrapperPage;
                             requestUri = wrapper?.WebTemplatePage.Data.Html;
                         }
+                        if (this.CurrentResponse == null)
+                        {
+                            Session.Current.Data = response.Resource as Json;
```
Comment (warpech) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: For me this looks smelly. An app can only have one object assigned to a session. In case of WebsiteProvider, session is already assigned [here](https://github.com/StarcounterApps/Website/blob/b712373da24bac96faba9faaba396c0726c74e37/src/WebsiteProvider/Api/ContentHandlers.cs#L63)

@chrhol can you pls review the commit https://github.com/StarcounterApps/Website/pull/65/commits/b712373da24bac96faba9faaba396c0726c74e37? It is supposed to fix PR #63

**Comment by warpech** [comment]
@chrhol ping, can you pls take a look at https://github.com/StarcounterApps/Website/pull/65#discussion_r108710678?

File: src/WebsiteProvider/Api/ContentHandlers.cs (hunk 2 above)
Comment (chrhol) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: I'm having a bit of a problem understanding exactly what these lines do. As @warpech states, the session is assigned from another place. Is it to ensure that it's always contains something?

Comment (warpech) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: @un-tone can you explain this line a bit? Or should we have a meeting?

**Comment by warpech** [comment]
@un-tone can you pls add tests?

**Comment by un-tone** [comment]
Tests are added.

File: src/WebsiteProvider/Api/ContentHandlers.cs (hunk 2 above)
Comment (un-tone) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: This is for case when WebsiteProvider doesn't wrap a response and returned it as-is. I assumed that in this case the Session.Data have to be changed. Otherwise, the responded JSON Patch doesn't contain the complete data and nothing is happened on the page.

File: test/WebsiteProvider_AcceptanceHelperOne/Helpers/DataHelper.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@ namespace WebsiteProvider_AcceptanceHelperOne
 {
     public class DataHelper
     {
-        public void GenerateData()
+        public void SetDefaultCatchingRules()
```
Comment (warpech) [inline test/WebsiteProvider_AcceptanceHelperOne/Helpers/DataHelper.cs]: It is hard to follow how does the catching rules config look like after running this helper.

I think that a helper that deletes all data and creates only the ones used in test would be better. WDYT?

File: src/WebsiteProvider/Api/ContentHandlers.cs (hunk 2 above)
Comment (warpech) [inline src/WebsiteProvider/Api/ContentHandlers.cs]: Ok so this will be long. The test `GoToOwnPage_NoCatchAllRules_LoadedAsIs` tests:

1. Loading `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/SetupNoCatchAllRuleTest` to prepare the test setup
2. Going to `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne` - a page that is stateless (does not create a session), but is configured to be wrapped in a surface by WebsiteProvider. WebsiteProvider creates the session.
3. Clicking on a "simple page" link that goes to `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/SimplePage` - a page that is also stateless and configured to be NOT wrapped in a surface by WebsiteProvider

What happens in lines https://github.com/StarcounterApps/Website/blob/8309d12b240a49f2f364d122d13851b423b1dbc9/src/WebsiteProvider/Api/ContentHandlers.cs#L139-L143 is that WebsiteProvider creates a session of response from a different app, and by some mechanism (which one?) makes the response replace the root of the previous response.

So here are some questions:

1. Should we actually wrap stateless pages? I think yes.
2. What should we do when morphing to a stateless page? Should we force it into the session? Or should we actually force a full page reload to that page? (I suppose the latter)
2. @un-tone please note that going to the page `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/SimplePage` directly actually does not render it. This is because it does not have a session, which causes it to be ignored by middleware (see [this line in middleware](https://github.com/Starcounter/level1/blob/2d6990034e47906453c8108c6191bdf43aa78959/src/Starcounter/Middleware/PartialToStandaloneHtmlProvider.cs#L97))

**Comment by warpech** [comment]
Continued discussion from https://github.com/StarcounterApps/Website/pull/65#discussion_r111356247:

-----

The test `GoToOwnPage_NoCatchAllRules_LoadedAsIs` tests:

1. Loading `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/SetupNoCatchAllRuleTest` to prepare the test setup
2. Going to `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne` - a page that is stateless (does not create a session), but is configured to be wrapped in a surface by WebsiteProvider. WebsiteProvider creates the session.
3. Clicking on a "simple page" link that goes to `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/SimplePage` - a page that is also stateless and configured to be NOT wrapped in a surface by WebsiteProvider

What happens in lines https://github.com/StarcounterApps/Website/blob/8309d12b240a49f2f364d122d13851b423b1dbc9/src/WebsiteProvider/Api/ContentHandlers.cs#L139-L143 is that WebsiteProvider creates a session of response from a different app, and by some mechanism (which one?) makes the response replace the root of the previous response.

@un-tone please note that going to the page `http://localhost:8080/WebsiteProvider_AcceptanceHelperOne/SimplePage` directly actually does not render it. This is because it does not have a session, which causes it to be ignored by middleware (see [this line in middleware](https://github.com/Starcounter/level1/blob/2d6990034e47906453c8108c6191bdf43aa78959/src/Starcounter/Middleware/PartialToStandaloneHtmlProvider.cs#L97))

So here are some questions:

1. Should we actually wrap stateless pages? I think yes.
3. What should we do when morphing to a stateless page? Should we force it into the session? Or should we actually force a full page reload to that page? (I suppose the latter)

@chrhol, @miyconst, WDYT?

**Comment by un-tone** [comment]
> The test GoToOwnPage_NoCatchAllRules_LoadedAsIs performs the following (@un-tone, right?)

Right.

File: test/WebsiteProvider_AcceptanceHelperOne/Api/MainHandlers.cs [hunk 4]
```diff
@@ -1,16 +1,44 @@
...
             Handle.GET("/WebsiteProvider_AcceptanceHelperOne/EmptyJson", () => new Json());
-            Handle.GET("/WebsiteProvider_AcceptanceHelperOne/EmptyPage",
-                () => new Page { Html = "/WebsiteProvider_AcceptanceHelperOne/Master.html" });
+            Handle.GET("/WebsiteProvider_AcceptanceHelperOne/SimplePage",
+                () => new Page { Html = "/WebsiteProvider_AcceptanceHelperOne/SimplePage.html" });
+
+            Handle.GET("/WebsiteProvider_AcceptanceHelperOne/SetDefaultCatchingRules", () =>
+            {
+                DataHelper.SetDefaultCatchingRules();
+                return Response.FromStatusCode(200);
```
Comment (un-tone) [inline test/WebsiteProvider_AcceptanceHelperOne/Api/MainHandlers.cs]: Fixed.

File: test/WebsiteProvider_AcceptanceHelperOne/Helpers/DataHelper.cs (hunk 3 above)
Comment (un-tone) [inline test/WebsiteProvider_AcceptanceHelperOne/Helpers/DataHelper.cs]: Agree. It makes tests dependent to the Website default data.

Comment (un-tone) [inline test/WebsiteProvider_AcceptanceHelperOne/Helpers/DataHelper.cs]: Done. Also added checkings for a loaded surface.

**Comment by un-tone** [comment]
I will add comments to the test.

**Comment by warpech** [comment]
Ping @miyconst, @chrhol regarding https://github.com/StarcounterApps/Website/pull/65#issuecomment-293850988

**Comment by miyconst** [comment]
Once again we are hit by out session mechanism. In other web techs that I am familiar is the session is created automatically, and you don't have to care about it. Still, there is an option to forcibly disable it, if needed.

Regarding the questions:

>Should we actually wrap stateless pages? I think yes.

I also think yes. If an app does not require session it still should be possible to wrap it with Website app.

>What should we do when morphing to a stateless page? Should we force it into the session? Or should we actually force a full page reload to that page? (I suppose the latter)

I don't see how we could actually check if we redirect to a session-state or session-less page. The only way to check, is to make a request when no session available, and check if session was created after the response.

Anyway, I think we should bring that page into current session, since the page is wrapped within Website app, and Website app requires session.

**Comment by warpech** [comment]
@chrhol what's your opinion on the discussion from https://github.com/StarcounterApps/Website/pull/65#discussion_r109866310 forward?

**Comment by chrhol** [comment]
I agree with @miyconst points for anwers to the questions. 
 
> I don't see how we could actually check if we redirect to a session-state or session-less page.

> Anyway, I think we should bring that page into current session, since the page is wrapped within Website app, and Website app requires session.

Since the website app requires session, everything wrapped in it should be brought in. 

Two additional points. 
- The code mentioned in a comment above (and is quoted underneath) is unsafe to use and there will be problems when more than one request is handled at the same time (I'm not sure that it belongs to this PR though).  Described in #72 
> What happens in lines https://github.com/StarcounterApps/Website/blob/8309d12b240a49f2f364d122d13851b423b1dbc9/src/WebsiteProvider/Api/ContentHandlers.cs#L139-L143 is that WebsiteProvider creates a session of response from a different app, and by some mechanism (which one?) makes the response replace the root of the previous response.
- Hopefully we can come to a decision on a way forward around how `Session` is working which probably will improve the code here for adding and storing viewmodels on the session (https://github.com/Starcounter/level1/issues/4159)

**Comment by un-tone** [comment]
> The code mentioned in a comment above (and is quoted underneath) is unsafe to use

Replaced `CurrentResponse` from the condition with local variable

**Comment by warpech** [comment]
Thanks @un-tone! I believe that `CurrentResponse` and `CurrentTemplate` are code smells and should be refactored to function parameters or view-model properties. Would you have some idea how to improve that?

----

Regarding the main topic of the doubtful lines https://github.com/StarcounterApps/Website/blob/8309d12b240a49f2f364d122d13851b423b1dbc9/src/WebsiteProvider/Api/ContentHandlers.cs#L139-L143, I propose to change them into:

```cs
                        if (!isWrapped)
                        {
                            //Morph to a view that is stateless and not catched by any surface
                            //This is tested by RedirectToOtherAppPageTest
                            //Should be improved by https://github.com/Starcounter/level1/issues/4159
                            Session.Current.Data = response.Resource as Json;
                        }
```

Notice that I have removed the following part:

```cs
                            if (Session.Current == null)
                            {
                                Session.Current = new Session(SessionOptions.PatchVersioning);
                            }
```

My logic behind this is that if we are morphing **out**, the session will always exist. All the tests are still passing. @un-tone, any flaws in my thinking here?

**Comment by un-tone** [comment]
Please review last changes.

File: src/WebsiteProvider/Helpers/Storage.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
+﻿using System;
+using System.Collections.Concurrent;
+
+namespace WebsiteProvider
+{
+    public class Storage<T>
+    {
+        private readonly ConcurrentDictionary<string, T> responses = new ConcurrentDictionary<string, T>();
```
Comment (warpech) [inline src/WebsiteProvider/Helpers/Storage.cs]: What are the benefits of `ConcurrentDictionary` vs a `Dictionary`? Should we always use `ConcurrentDictionary`? Then let's add it to our guidelines.
